Railed Move Blocks - Attached Above, and Eased Vertical Movement additions

MorkyPep opened the pull request:
This addition will add two features to Railed Move Blocks.

Attached Above - clicking this option will allow static movers to render above the railed move block steering buttons.

Eased Vertical Movement - clicking this option will allow diagonal inputs to more easily move the railed move block vertically. This is mainly for players on analog. My upcoming map uses a new move/mechanic that requires the player to move the railed move blocks with diagonal inputs. It was very tight and clunky on analog. This code fixes the problem. DM me if you want to see what it is used for, I can send a video. 

This is my first ever PR, so I apologize if I did something incorrectly. I plan to release my map once these changes are live!
DM on discord for any questions/concerns

morkypep50#5249
Closes #132.

MorkyPep committed:
Update RailedMoveBlock.cs 

Added two options:

- Render Above will render static movers above the railed move block steering buttons

- Eased Vertical Movement will allow diagonal inputs to more easily move the railed move block vertically. This is used to make certain movement easier on analog.

MorkyPep committed:
Update railed_move_block.lua

Added plugin options for Attached Above and Eased Vertical Movement

MorkyPep committed:
Update en_gb.lang

added descriptions for Attached Above and Eased Vertical Movement

MorkyPep committed:
Update RailedMoveBlock.cs

mayromr commented on Loenn/entities/misc/railed_move_block.lua:
```diff
@@ -35,11 +35,14 @@ for i, steeringMode in ipairs(steeringModes) do
-            speed = 120.0
+            speed = 120.0,
+			easedVerticalMovement = false,
+			attachedAbove = false
```
```suggestion
            speed = 120.0,
            easedVerticalMovement = false,
            attachedAbove = false
```

Tiny nit, we use spaces to indent

MorkyPep replied:
fixed.

mayromr commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -254,6 +260,15 @@ public override void Awake(Scene scene)
+            foreach (StaticMover staticMover in this.staticMovers)
+            {
+                staticMover.Entity.Depth = this.Depth - 1;
+            }
```
```suggestion
            foreach (StaticMover staticMover in staticMovers)
            {
                staticMover.Entity.Depth = Depth - 1;
            }
```
Tiny nit, `this` is redundent here

MorkyPep replied:
fixed

mayromr commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -308,22 +323,38 @@ public override void Update()
+                {
+                    newSpeed = moveSpeed * Math.Sign(Input.Feather.Value.Y) * (dir.Y > 0f ? 1f : -1f);
+                    newFillColor = MoveBgFill;
+                    icon = Input.MoveY.Value == 1 ? DownIcon : UpIcon;
```
```suggestion
                    icon = Math.Sign(Input.Feather.Value.Y) == 1 ? DownIcon : UpIcon;
```

We should use the sign of `Input.Feather.Value.Y` to determine the direction the player is moving in, because we could go down using using the value from `Input.Feather.Value.Y` while `Input.MoveY.Value`  is still zero.

MorkyPep replied:
fixed and tested

mayromr commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -308,22 +323,38 @@ public override void Update()
                 newSpeed = moveSpeed * Input.MoveX.Value * (dir.X > 0f ? 1f : -1f);
                 newFillColor = MoveBgFill;
                 icon = Input.MoveX.Value == 1 ? RightIcon : LeftIcon;
             }
```
Small nit, but I feel like it would be easier to read this if statement if it was restructured as
```csharp
if((topPressed || bottomPressed) && HasPlayerRider())
{
    if (easedVerticalMovement && (Input.Feather.Value.Y > 0.4f || Input.Feather.Value.Y < -0.4f))
    {
        // ...
    }
    else if (Input.MoveY.Value != 0)
    {
        // ...
    }
} else if ((leftPressed || rightPressed) && HasPlayerClimbing() && Input.MoveX.Value != 0)
{
    // ...
}
```

MorkyPep replied:
fixed and tested! sorry it took so long to make these changes. I've been sick for the past few weeks and just didn't have the willpower to work on my project. I really appreciate your feedback and I've updated the branch with your suggested changes as well as tested them to make sure everything is working correctly.

mayromr replied:
No worries, no one works here you can take as much time as you need

mayromr reviewed (commented):
Thanks for contribution! this PR looks great, I left a few comments I would appreciate if you could take a look at them.

MorkyPep committed:
Update railed_move_block.lua

slight formatting change

MorkyPep committed:
Update RailedMoveBlock.cs

minor structure adjustments plus fix on block icon not properly displaying downwards.

mayromr commented:
LGTM 👍

catapillie commented on Loenn/lang/en_gb.lang:
```diff
@@ -566,6 +566,8 @@ entities.CommunalHelper/RailedMoveBlock.placements.name.vertical=Railed Move Blo
 entities.CommunalHelper/RailedMoveBlock.attributes.description.steeringMode=Represents the way in which you can steer the block using the side buttons.\n\n- "Horizontal" will enable buttons on the left and right side of the block.\n- "Vertical" will enable one button at the top of the block.\n- "Both" will enable the two above.
 entities.CommunalHelper/RailedMoveBlock.attributes.description.speed=The maximum speed this block can go at.
+entities.CommunalHelper/RailedMoveBlock.attributes.description.easedVerticalMovement=If checked, diagonal inputs will more easily move the Railed Move Block vertically. This is used to make some moves easier on analog.
+entities.CommunalHelper/RailedMoveBlock.attributes.description.attachedAbove=If checked, attached entities will be rendered above the steering buttons. (Springs,spikes etc.) 
```
```suggestion
entities.CommunalHelper/RailedMoveBlock.attributes.description.attachedAbove=If checked, attached entities will be rendered above the steering buttons. (Springs, spikes, etc.) 
```
Punctuation

catapillie commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -254,6 +260,15 @@ public override void Awake(Scene scene)
 
         scene.Add(border = new Border(this));
         scene.Add(pathRenderer = new RailedMoveBlockPathRenderer(this));
+
```
```suggestion
```
another extra line you can remove

catapillie commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -233,6 +238,7 @@ public RailedMoveBlock(Vector2 position, int width, int height, Vector2 node, St
         });
         sfx.Play(CustomSFX.game_railedMoveBlock_railedmoveblock_move, "arrow_stop", 1f);
         Add(new LightOcclude(0.5f));
+
```
```suggestion
```
formatting nitpick: this line can be removed

catapillie commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -308,22 +323,36 @@ public override void Update()
+                    icon = Input.MoveY.Value == 1 ? DownIcon : UpIcon;
+                }
             }
+
```
```suggestion
```
formatting

catapillie commented on src/Entities/Misc/RailedMoveBlock.cs:
```diff
@@ -308,22 +323,36 @@ public override void Update()
                 icon = Input.MoveX.Value == 1 ? RightIcon : LeftIcon;
             }
 
+        
+
```
```suggestion
```
more formatting

catapillie reviewed (commented):
Looks good, just some formatting and punctuation nitpicks.

catapillie committed:
fix punctuation

catapillie committed:
formatting: remove extra empty lines

catapillie commented:
Took care of it. Thanks, merging.
